Update release notes (#4996)
kedro-org/kedro: RELEASE.md

# Upcoming Release 1.0.0
## Major features and improvements

### DataCatalog

> * Previously experimental `KedroDataCatalog` has now been renamed as `DataCatalog` and has replaced the previous implementation as the default catalog class. It now supports a dict-like interface, lazy dataset initialization, and has improved performance. The users of Kedro as a framework should notice no difference in their workflows, however the new `DataCatalog` offers new API when used programatically. Read more about it in the [Kedro documentation](https://docs.kedro.org/en/unreleased/catalog-data/advanced_data_catalog_usage/).

Reviewer: `stable` for the doc link

> * Previously experimental `KedroDataCatalog` has now been renamed as `DataCatalog` and has replaced the previous implementation as the default catalog class. It now supports a dict-like interface, lazy dataset initialization, and has improved performance. The users of Kedro as a framework should notice no difference in their workflows, however the new `DataCatalog` offers new API when used programatically. Read more about it in the [Kedro documentation](https://docs.kedro.org/en/unreleased/catalog-data/advanced_data_catalog_usage/).

Reviewer: `stable` link here instead of `unreleased`

> * Previously experimental `KedroDataCatalog` has now been renamed as `DataCatalog` and has replaced the previous implementation as the default catalog class. It now supports a dict-like interface, lazy dataset initialization, and has improved performance. The users of Kedro as a framework should notice no difference in their workflows, however the new `DataCatalog` offers new API when used programatically. Read more about it in the [Kedro documentation](https://docs.kedro.org/en/unreleased/catalog-data/advanced_data_catalog_usage/).

Reviewer 2: Yep although they will break until `stable` is in fact released

* Removed `kedro catalog create` CLI command.
* Removed the following `DataCatalog` methods: `_get_dataset()`, `add_all()`, `add_feed_dict()`, `list()`, and `shallow_copy()`.

### Other features and improvements
* Added stricter validation to dataset names in the `Node` class, ensuring `.` characters are reserved to be used as part of a namespace.
* Added a `prefix_datasets_with_namespace` argument to the `Pipeline` class which allows users to turn on or off the prefixing of the namespace to the node inputs, outputs, and parameters.
* Added support for running multiple namespaces within a single session with `--namespaces` CLI option.
* Improved namespace validation efficiency to prevent significant slowdowns when creating large pipelines.
* Changed the default node name to be formed of the function name used in the node suffixed by a secure hash (SHA-256) based on the function, inputs, and outputs, ensuring uniqueness and improved readability.

> * Added an option to select which multiprocessing start method is going to be used on `ParallelRunner` via the `KEDRO_MP_CONTEXT` environment variable.

Author: ```suggestion
* Added an option to select which multiprocessing start method is going to be used on `ParallelRunner` via the `KEDRO_MP_CONTEXT` environment variable.
* Added `--only-missing-outputs` CLI flag to `kedro run`. This flag skips nodes when all their persistent outputs exist.
* Removed the `AbstractRunner.run_only_missing()` method, an older and underused API for partial runs. Please use `--only-missing-outputs` CLI instead.
```

* Renamed instances of `extra_params` and `_extra_params` to `runtime_params`.
* Removed the `modular_pipeline` module and moved functionality to the `pipeline` module instead.
* Renamed `ModularPipelineError` to `PipelineError`.
* `Pipeline.grouped_nodes_by_namespace()` was replaced with `group_nodes_by(group_by)`, which supports multiple strategies and returns a list of `GroupedNodes`, improving type safety and consistency for deployment plugin integrations.
* The micro-packaging feature and the corresponding `micropkg` CLI command have been removed.
* Renamed `session_id` parameter to `run_id` in all runner methods and hooks to improve API clarity and prepare for future multi-run session support.
* Removed the following `DataCatalog` methods: `_get_dataset()`, `add_all()`, `add_feed_dict()`, `list()`, and `shallow_copy()`.
* Removed the CLI command `kedro catalog create`.
* Changed the output of `runner.run()` — it now always returns all pipeline outputs, regardless of catalog configuration.

## Documentation changes
* Revamped the look and feel of the Kedro documentation, including a new theme and improved navigation with `mkdocs` as the documentation engine.

> * Updated the `DataCatalog` documentation with improved structure and detailed description of new features. Read the [DataCatalog documentation here](https://docs.kedro.org/en/unreleased/catalog-data/introduction/).

Reviewer: this link will need to be updated to `https://docs.kedro.org/en/stable/catalog-data/introduction/`

## Community contributions
Many thanks to the following Kedroids for contributing PRs to this release:
* [Yury Fedotov](https://github.com/yury-fedotov)
* [Kitsios Konstantinos](https://github.com/kitsiosk)

## Migration guide from Kedro 0.19.* to 1.*

> [See the migration guide for 1.0.0 in the Kedro documentation](https://docs.kedro.org/en/unreleased/about/migration/).

Reviewer: this should be `stable` as well

# Upcoming Release 1.0.0
## Major features and improvements

> ### DataCatalog

Reviewer 3: ```suggestion
### Data Catalog
```

> * Previously experimental `KedroDataCatalog` has now been renamed as `DataCatalog` and has replaced the previous implementation as the default catalog class. It now supports a dict-like interface, lazy dataset initialisation, and has improved performance. The users of Kedro as a framework should notice no difference in their workflows, however, the new `DataCatalog` offers new API when used programatically. Read more about it in the [Kedro documentation](https://docs.kedro.org/en/stable/catalog-data/advanced_data_catalog_usage/).

Reviewer 4: ```suggestion
* Previously experimental `KedroDataCatalog` has now been renamed to `DataCatalog` and has replaced the previous implementation as the default catalog class. It now supports a dict-like interface, lazy dataset initialisation, and improved performance. While users of Kedro as a framework should notice no difference in their workflows, the new `DataCatalog` offers an extended API when used programmatically. This includes support for powerful pipeline-based commands (available via both CLI and interactive environments), simplified usage of dataset factories, centralised pattern resolution through the new `CatalogConfigResolver` property, and the ability to serialise and reconstruct catalogs from configuration.  Read more about it in the [Kedro documentation](https://docs.kedro.org/en/stable/catalog-data/advanced_data_catalog_usage/).
```

> * Previously experimental `KedroDataCatalog` has now been renamed to `DataCatalog` and has replaced the previous implementation as the default catalog class. It now supports a dict-like interface, lazy dataset initialisation, and improved performance. While users of Kedro as a framework should notice no difference in their workflows, the new `DataCatalog` offers an extended API when used programmatically. This includes support for powerful pipeline-based commands (available via both CLI and interactive environments), simplified usage of dataset factories, centralised pattern resolution through the new `CatalogConfigResolver` property, and the ability to serialise and reconstruct catalogs from configuration.  Read more about it in the [Kedro documentation](https://docs.kedro.org/en/stable/catalog-data/advanced_data_catalog_usage/).

Reviewer 3: ```suggestion
* Renamed `KedroDataCatalog` to `DataCatalog`, replacing the previous implementation of the default catalog class. Previously an experimental feature, it supports a dict-like interface, lazy dataset initialisation, and improved performance. While users of Kedro as a framework should notice no difference in their workflows, the new `DataCatalog` offers an extended API when used programmatically. This includes support for powerful pipeline-based commands (available via both CLI and interactive environments), simplified usage of dataset factories, centralised pattern resolution through the new `CatalogConfigResolver` property, and the ability to serialise and reconstruct catalogs from configuration.  Read more about it in the [Kedro documentation](https://docs.kedro.org/en/stable/catalog-data/advanced_data_catalog_usage/).
```

Reworded the first sentence to be more clear and in line with other release note statements.

That said, I also think this is _really_ long for a release note entry and could be condensed/delegated to the documentation and/or migration guide. The in-depth benefits of the data catalog also seem like they should be covered in a 1.0.0 release blog more than buried in the release notes.

* Removed `pre-commit-hooks` dependency for new project creation.

## Breaking changes to the API
* Private methods `_is_project` and `_find_kedro_project` are changed to `is_kedro_project` and `find_kedro_project`.
* Renamed instances of `extra_params` and `_extra_params` to `runtime_params`.
* Removed the `modular_pipeline` module and moved functionality to the `pipeline` module instead.
* Renamed `ModularPipelineError` to `PipelineError`.
* `Pipeline.grouped_nodes_by_namespace()` was replaced with `group_nodes_by(group_by)`, which supports multiple strategies and returns a list of `GroupedNodes`, improving type safety and consistency for deployment plugin integrations.
* The micro-packaging feature and the corresponding `micropkg` CLI command have been removed.
* Renamed `session_id` parameter to `run_id` in all runner methods and hooks to improve API clarity and prepare for future multi-run session support.
* Removed the following `DataCatalog` methods: `_get_dataset()`, `add_all()`, `add_feed_dict()`, `list()`, and `shallow_copy()`.
* Removed the CLI command `kedro catalog create`.

> * Changed the output of `runner.run()` — it now always returns all pipeline outputs, regardless of catalog configuration.

Reviewer 4: Shall we also mention here that `session.run()` output has changed as well?

# Upcoming Release 1.0.0
## Major features and improvements
## Bug fixes and other changes
## Breaking changes to the API
## Upcoming deprecations for Kedro 1.0.0
## Documentation changes
## Community contributions

# Release 1.0.0rc3

## Major features and improvements

> Changed `DataCatalog.__getitem__` to raise `DatasetNotFoundError` for missing datasets, aligning with expected dictionary behavior.

Author: @SajidAlamQB I didn't include this because it was a bug fix in a new API between RC releases so it shouldn't be mentioned

### Other features and improvements
* Changed the default node name to be formed of the function name used in the node suffixed by a secure hash (SHA-256) based on the function, inputs, and outputs, ensuring uniqueness and improved readability.
* Added an option to select which multiprocessing start method is going to be used on `ParallelRunner` via the `KEDRO_MP_CONTEXT` environment variable.

> * Added `--only-missing-outputs` CLI flag to `kedro run`. This flag skips nodes when all their persistent outputs exist.

Author: It is under the "Major features and improvements" section I just made sub sections for Datacatalog and Namespaces separately

> * Removed the `AbstractRunner.run_only_missing()` method, an older and underused API for partial runs. Please use `--only-missing-outputs` CLI instead.

Author: Done

# Upcoming Release 1.0.0
## Major features and improvements

### Data Catalog

> * Renamed `KedroDataCatalog` to `DataCatalog`, replacing the previous implementation of the default catalog class. Previously an experimental feature, it supports a dict-like interface, lazy dataset initialisation, and improved performance. While users of Kedro as a framework should notice no difference in their workflows, the new `DataCatalog` offers an extended API when used programmatically. This includes support for powerful pipeline-based commands (available via both CLI and interactive environments), simplified usage of dataset factories, centralised pattern resolution through the new `CatalogConfigResolver` property, and the ability to serialise and reconstruct catalogs from configuration.  Read more about it in the [Kedro documentation](https://docs.kedro.org/en/stable/catalog-data/advanced_data_catalog_usage/).

Reviewer 4: Maybe something like? I can add more/less features to the list if needed.

```suggestion
* The previously experimental `KedroDataCatalog` has been renamed to `DataCatalog` and is now the default catalog implementation.
* It retains the dict-like interface, supports lazy dataset initialisation, and delivers improved performance.
* While this change is seamless for users following standard Kedro workflows, it introduces a richer API for programmatic use:
  * New pipeline-aware commands, available via both the CLI and interactive environments.
  * Simplified handling of dataset factories.
  * Centralised pattern resolution via the `CatalogConfigResolver` property.
  * Ability to serialise the catalog to configuration and reconstruct it from it.

Read more in the [Kedro documentation](https://docs.kedro.org/en/stable/catalog-data/advanced_data_catalog_usage/).

```

Reviewer 3: I just think the amount of content on this is a lot for release notes, whether it's broken down into more bullets or not (I think more bullets is worse actually). The main thing to note in 1.0.0 is that `KedroDataCatalog` has been renamed (if you were using it), and the default Data Catalog implementation changed.

As written, for most users this may not be something they think about, and so 1.0.0 release notes can highlight it yet. Still think pointing to or later creating content diving into benefits of new implementation, more in prose (e.g. "What's new in 1.0.0?" section of release blog) could be more effective.

> # Upcoming Release 1.0.0

Reviewer 5: [Nit] will this be changed to Release 1.0.0 in the PR - https://github.com/kedro-org/kedro/pull/5000/files ?
